### PR TITLE
Migrate to Tailwind

### PR DESCRIPTION
**Pull request by paulglx**
Migrate the frontend from Bootstrap to TailwindCSS

### CODE REVIEW

File: frontend/src/components/Course/CourseComponents/SessionContent.tsx [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+                    <span className="text-gray-700 text-xl font-medium mb-2">{session.title}</span>
+                        <div className="flex items-center">
+                            <ClockIcon className="w-6 h-6 mr-1 text-gray-500" />
+                            <span className="text-gray-500 text-sm">{session.duration}</span>
+                                {session.openingTime && (
+                            <div className="flex items-center ml-2">
+                            <LockOpenIcon className="w-5 h-5 mr-1 text-gray-500" />
+                                <span className="text-gray-500 text-sm">{session.openingTime}</span>
+                            </div>
+                            )}
```
Comment (paulglx) [inline frontend/src/components/Course/CourseComponents/SessionContent.tsx:93]: There are no openingTime and duration fields on the session model

File: frontend/src/components/Exercise/ExerciseRuntimeTab.tsx [hunk 2]
```diff
@@ -1,25 +1,20 @@
...
 const ExerciseRuntimeTab = (props: any) => {
 
     const { course, session, exercise, isOwner } = props
 
     const [prefix, setPrefix] = useState("");
     const [suffix, setSuffix] = useState("");
-    const dispatch = useDispatch();
     const [updateExercise] = useUpdateExerciseMutation();
+    const [notificationMessage, setNotificationMessage] = useState("");
+    const [notificationType, setNotificationType] = useState("");
```
Comment (paulglx) [inline frontend/src/components/Exercise/ExerciseRuntimeTab.tsx:15]: There exists a reusable notification system, used like :

```
dispatch(pushNotification({
    message: "...",
    type: "success" / "error" / ...
})
```

It has been updated, the types used to correspond to Bootstrap colors.

We can keep using it.

File: frontend/src/components/Exercise/ExerciseRuntimeTab.tsx [hunk 3]
```diff
@@ -1,25 +1,20 @@
...
+    const [isNotificationVisible, setIsNotificationVisible] = useState(false);
 
-    useEffect(() => {
-        setPrefix(exercise.prefix);
-        setSuffix(exercise.suffix);
-    }, [exercise])
```
Comment (paulglx) [inline frontend/src/components/Exercise/ExerciseRuntimeTab.tsx:22]: Was this unnecessary ?

File: frontend/src/components/Exercise/ExerciseRuntimeTab.tsx [hunk 4]
```diff
@@ -28,92 +23,139 @@ const ExerciseRuntimeTab = (props: any) => {
...
+                    <button onClick={handleCloseNotification}>Close</button>
+                </div>
+            )}
+        </div>
+
+        <div className="border rounded-lg bg-white shadow-md p-4 mb-4">
+            <h5 className="text-2xl font-bold mb-2">Code preview</h5>
+            <p className="text-gray-600 mb-4">This is what the tested file will look like.</p>
+            <pre className="border rounded-lg bg-gray-100 p-4">
+                {prefix && (
```
Comment (paulglx) [inline frontend/src/components/Exercise/ExerciseRuntimeTab.tsx:138]: If the prefix is empty, React will render an empty string as `""` evaluates to `false` in js.
Maybe use safer and more explicit ternary operator `expr ? foo : bar` ?

See [https://kentcdodds.com/blog/use-ternaries-rather-than-and-and-in-jsx](https://kentcdodds.com/blog/use-ternaries-rather-than-and-and-in-jsx)

File: frontend/src/components/Exercise/ExerciseTestsTab.tsx [hunk 5]
```diff
@@ -67,101 +67,177 @@ const ExerciseTestsTab = (props: any) => {
...
-                                    value={test?.stdin}
+            {exerciseIsSuccess && tests && (
+                <>
+                    {tests.length > 0 ? (
+                        <h6 className="text-muted">
+                            {isOwner ? "Click test to edit" : "Tests can be edited by owners"}
+                        </h6>
+                    ) : (
+                        <div className="bg-gray-100 shadow-lg shadow-gray-400/70 p-6 md:p-2 flex justify-center items-center flex-col rounded-lg">
+                            <h6 className="text-muted p-2 md:p-4">This course doesn't have any test.</h6>
```
Comment (paulglx) [inline frontend/src/components/Exercise/ExerciseTestsTab.tsx:80]: "This _session_ doesn't have any tests"

File: frontend/src/components/Exercise/ExerciseRuntimeTab.tsx (hunk 3 above)
Comment (sepulvedalili) [reply frontend/src/components/Exercise/ExerciseRuntimeTab.tsx:22]: Yes, this was unnecessary.

File: frontend/src/components/Exercise/ExerciseRuntimeTab.tsx (hunk 2 above)
Comment (sepulvedalili) [reply frontend/src/components/Exercise/ExerciseRuntimeTab.tsx]: Done

File: frontend/src/components/Exercise/ExerciseRuntimeTab.tsx (hunk 4 above)
Comment (sepulvedalili) [reply frontend/src/components/Exercise/ExerciseRuntimeTab.tsx]: Done

File: frontend/src/components/Exercise/ExerciseTestsTab.tsx (hunk 5 above)
Comment (sepulvedalili) [reply frontend/src/components/Exercise/ExerciseTestsTab.tsx]: Done

File: frontend/src/components/Exercise/ExerciseRuntimeTab.tsx (hunk 3 above)
Comment (sepulvedalili) [reply frontend/src/components/Exercise/ExerciseRuntimeTab.tsx]: I have reverted my changes

**Review by sepulvedalili — APPROVED** [review]
LGTM

**Pull request merged**